File: lib/scanner/score.ts
import type { Finding, CategoryScore, Category } from "./types";
import { CATEGORY_LABELS } from "./types";

export function gradeFromScore(pct: number): string {
  if (pct >= 90) return "A";
  if (pct >= 80) return "B";
  if (pct >= 70) return "C";
  if (pct >= 55) return "D";
  return "F";
}

const CATEGORY_ORDER: Category[] = ["transport", "headers", "dns", "cookies", "disclosure"];

export function scoreByCategory(findings: Finding[]): {
  categories: CategoryScore[];
  overall: number;
} {
  const categories: CategoryScore[] = [];
  let totalEarned = 0;
  let totalMax = 0;

  for (const cat of CATEGORY_ORDER) {
    const items = findings.filter((f) => f.category === cat && f.maxPoints > 0);
    if (items.length === 0) continue;
    const earned = items.reduce((a, f) => a + f.points, 0);
    const max = items.reduce((a, f) => a + f.maxPoints, 0);
    const all = findings.filter((f) => f.category === cat);
    categories.push({
      category: cat,
      label: CATEGORY_LABELS[cat],
      score: max > 0 ? Math.round((earned / max) * 100) : 0,
      earned,
      max,
      pass: all.filter((f) => f.status === "pass").length,
      warn: all.filter((f) => f.status === "warn").length,
      fail: all.filter((f) => f.status === "fail").length,
    });
    totalEarned += earned;
    totalMax += max;
  }

  const overall = totalMax > 0 ? Math.round((totalEarned / totalMax) * 100) : 0;
  return { categories, overall };
}

const STATUS_ORDER: Record<Finding["status"], number> = { fail: 0, warn: 1, pass: 2, info: 3 };
const SEVERITY_ORDER: Record<Finding["severity"], number> = {
  critical: 0,
  high: 1,
  medium: 2,
  low: 3,
  info: 4,
};

export function sortFindings(findings: Finding[]): Finding[] {
  return [...findings].sort(
    (a, b) =>
      STATUS_ORDER[a.status] - STATUS_ORDER[b.status] ||
      SEVERITY_ORDER[a.severity] - SEVERITY_ORDER[b.severity] ||
      b.maxPoints - a.maxPoints
  );
}

File: engine/internal/scan/score.go
package scan

import "sort"

// GradeFromScore maps a 0–100 score to a letter grade.
func GradeFromScore(pct int) string {
	switch {
	case pct >= 90:
		return "A"
	case pct >= 80:
		return "B"
	case pct >= 70:
		return "C"
	case pct >= 55:
		return "D"
	default:
		return "F"
	}
}

// categoryOrder controls the display order of categories in a report. Unknown
// categories are appended after these, alphabetically.
var categoryOrder = map[Category]int{
	CategoryTransport:  0,
	CategoryHeaders:    1,
	CategoryDNS:        2,
	CategoryCookies:    3,
	CategoryContent:    4,
	CategoryDisclosure: 5,
	CategorySurface:    6,
	CategoryIntel:      7,
}

// scoreByCategory aggregates findings into per-category scores and an overall
// score. Only findings with MaxPoints > 0 contribute to scoring; informational
// findings still count toward pass/warn/fail tallies.
func scoreByCategory(findings []Finding) ([]CategoryScore, int) {
	type acc struct {
		earned, max          int
		pass, warn, fail     int
		hasScored            bool
	}
	buckets := map[Category]*acc{}
	for _, f := range findings {
		a := buckets[f.Category]
		if a == nil {
			a = &acc{}
			buckets[f.Category] = a
		}
		if f.MaxPoints > 0 {
			a.earned += f.Points
			a.max += f.MaxPoints
			a.hasScored = true
		}
		switch f.Status {
		case StatusPass:
			a.pass++
		case StatusWarn:
			a.warn++
		case StatusFail:
			a.fail++
		}
	}

	cats := make([]CategoryScore, 0, len(buckets))
	var totalEarned, totalMax int
	for cat, a := range buckets {
		if !a.hasScored && a.pass == 0 && a.warn == 0 && a.fail == 0 {
			continue
		}
		label := CategoryLabels[cat]
		if label == "" {
			label = string(cat)
		}
		score := 0
		if a.max > 0 {
			score = int(float64(a.earned)/float64(a.max)*100 + 0.5)
		}
		cats = append(cats, CategoryScore{
			Category: cat, Label: label, Score: score,
			Earned: a.earned, Max: a.max,
			Pass: a.pass, Warn: a.warn, Fail: a.fail,
		})
		totalEarned += a.earned
		totalMax += a.max
	}

	sort.Slice(cats, func(i, j int) bool {
		oi, iok := categoryOrder[cats[i].Category]
		oj, jok := categoryOrder[cats[j].Category]
		if iok && jok {
			return oi < oj
		}
		if iok != jok {
			return iok // known categories first
		}
		return cats[i].Category < cats[j].Category
	})

	overall := 0
	if totalMax > 0 {
		overall = int(float64(totalEarned)/float64(totalMax)*100 + 0.5)
	}
	return cats, overall
}

var statusRank = map[Status]int{StatusFail: 0, StatusWarn: 1, StatusPass: 2, StatusInfo: 3}
var severityRank = map[Severity]int{
	SeverityCritical: 0, SeverityHigh: 1, SeverityMedium: 2, SeverityLow: 3, SeverityInfo: 4,
}

// sortFindings orders findings worst-first: failures before warnings, then by
// severity, then by weight.
func sortFindings(findings []Finding) {
	sort.SliceStable(findings, func(i, j int) bool {
		a, b := findings[i], findings[j]
		if statusRank[a.Status] != statusRank[b.Status] {
			return statusRank[a.Status] < statusRank[b.Status]
		}
		if severityRank[a.Severity] != severityRank[b.Severity] {
			return severityRank[a.Severity] < severityRank[b.Severity]
		}
		return a.MaxPoints > b.MaxPoints
	})
}
